### DISCUSSION TITLE
Improving the Desktop Front End

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#506] [Design & Development]
Yesterday I held an introductory session on MsPASS to a group of first-year graduate students.   The session reinforced something I had been thinking for some time.  In a nutshell, our front-end for interactive use with MsPASS is awful and unquestionably an excessive energy barrier for 99% of potential users.  

Why?
1.  The cut-and-paste url approach to connecting to the container is a ridiculous system-dependent complexity that leaves people bewildered.  
2. Jupyter is currently a horrible platform for development.   We need to all realize most mortals aren't as proficient at programming as this team and need an environment can give them more feedback than a crash of a code box and the printout of the exception chain.   That is exceptionally true of junior graduate students like this group.   They were actually better than most as they had all completed a data analysis course the previous semester that was python based, but they were still rookies.  A strength of using python as the job control language in mspass is it is a powerful language but "powerful" also == complexity.   (A type example is C++.  It is a very powerful language than can do almost anything, but it is the most complex common computer language today.)

What do we need?
1.  I think we have a path to development for a better front end for HPC and cluster systems in the SCOPED gateway   It is the desktop (all-in-one container) that needs a different approach.
2. We need a more user-friendly front-end for desktop use designed on development.  By that I mean a flexible debugging environment and easier to startup.  

Let me expand point 2 above as an initial set of possible functional specifications for what this interface should do.  I suggest we use this forum in combination with our weekly meetings to turn this into a design document.

*Desktop Front-end functional specifications*

- Need a simple one line launcher that should normally just work.   
    - Probably requires a configuration file that may need to be edited before launcher is run.
    - Could also be used as the basis for a double-click startup to match standard app model of today

- Needs multiple, configurable options for the front end
   - Existing Jupyter lab OR notebook interface
   - spyder or some other standard IDE
   - unix shell running in the container (basis for a simple ipython interface)
 
*Possible solutions*
- This problem is probably solved if the current effort to set up MsPASS to build as an anaconda package is solved.   Anaconda includes spyder and its own launcher.  
- We discussed a simple shell/ipython interface in recent meetings.  Seems that can be done with a fairly minor change to the container internal launch script.
- I have seen online sources discuss how to setup a container's networking to interface with an ide like spyder.   That is a more complex problem from previous reading on this topic I did some time ago, but seems tractable.  There may be simpler solutions for other ide's but it is hard to say.  

Group, please contribute to this discussion and we should add it to the agenda for our weekly meetings until this problem is solved.  It is a fundamental usage barrier for MsPASS.

**Comment by wangyinz** [comment]
Thanks for sharing the experience! I guess it would be nice if we have something like how Matlab runs where a double-click brings up everything. But, this is almost impossible within a container environment as a GUI interface will need access and be compatible with the host OS, so it will be very easy to break even if we have something hacked together. Within the container, I guess the only thing doable is to switch to a pure command line interface, where the frontend will be just the Python prompt. Note that I prefers the basic Python over iPython for this because the add features of iPython might just be another layer of complexity for a beginner. If some advanced features are needed, it is probably a good idea to switch to Jupyter anyway. 

Now, if we get rid of the container, all other options are then possible. For IDE integration, I don't know much about that, but I can see we either tie to a specific one or trying to be compatible with any IDEs. I can see obvious downsides of each. The other option is to make a binary release of MsPASS (like an app such as Matlab), and we need to make it compatible with Windows, Mac OS and Linux. Not sure how difficult that would be, but we need to learn more about it as this is something I have no experience with before.

With all these said, the problem here is exactly why I pushed for the science gateway. Ideally, users can just go to the URL and run things without any of the setup process. Still, with the backend being HPC, it is still a bit complicated even with a gateway. The alternative is to use the cloud. There are services running on the cloud that is similar to what science gateway is. Some well-known ones are [Binder](https://mybinder.org/) and [JupyterHub](https://jupyter.org/hub), but I think they all come with a cost for serious users. We can given them a try I guess.

**Reply by pavlis** [reply]
Indeed the gateway is the solution for clusters of any kind and what we definitely need.  Also, seems much of the getting started problem can be solved if we can make the conda package work.   

One tidget is I did discover [this github page](https://github.com/timcera/spyder-desktop-ubuntu).  Not useful directly, but is suggest it might be possible to set up a container with support for spyder with some more additions.   @wangyinz  you might look at the dockerfile for that package and see how feasible that might be.  That site is a bit troubling as it is pretty old and spyder has changed a lot in that time period.
